### PR TITLE
fix(authn): session authn is skipped when anonymous policy is configured

### PR DESCRIPTION
**Pull request by peusebiu**
closes: #1642

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/authn.go [hunk 1]
```diff
@@ -394,22 +353,65 @@ func (amw *AuthnMiddleware) TryAuthnHandlers(ctlr *Controller) mux.MiddlewareFun
...
+					next.ServeHTTP(cloneResp, cloneReq)
+
+					return
+				}
+			} else {
+				// try anonymous auth only if basic auth/session was not given
+				// we want to bypass auth for mgmt route
+				isMgmtRequested := request.RequestURI == constants.FullMgmtPrefix
+				if ctlr.Config.HTTP.AccessControl.AnonymousPolicyExists() || isMgmtRequested {
+					// do we really need this? acCtx will be non nil!
```
Comment (andaaron) [inline pkg/api/authn.go:405]: Not sure what to say, I would keep the existing logic of creating the context, unless you also update all the places where it is checked for nil.

File: pkg/api/controller_test.go [hunk 2]
```diff
@@ -3559,151 +3559,190 @@ func TestAuthorizationWithBasicAuth(t *testing.T) {
...
+
+			client := resty.New()
+			client.SetBasicAuth(username, passphrase)
+
+			RunAuthorizationTests(t, client, baseURL, conf)
+		})
+	})
+}
+
+func TestGetUsername(t *testing.T) {
```
Comment (andaaron) [inline pkg/api/controller_test.go:3635]: Shouldn't we have a positive test besides the negative ones?

File: pkg/api/authn.go [hunk 3]
```diff
@@ -800,3 +803,39 @@ func hashUUID(uuid string) string {
...
+	// at this point we should have a session set on cookie.
+	// if created in the earlier Get() call then user is not logged in with sessions.
+	if session.IsNew {
+		return "", false
+	}
+
+	authenticated := session.Values["authStatus"]
+	if authenticated != true {
+		log.Error().Msg("can not get `user` session value")
+
```
Comment (peusebiu) [inline pkg/api/authn.go:829]: this should be covered by the existing tests, check what happened!

File: pkg/api/authn.go [hunk 4]
```diff
@@ -800,3 +803,39 @@ func hashUUID(uuid string) string {
...
+		return "", false
+	}
+
+	identity, ok := session.Values["user"].(string)
+	if !ok {
+		log.Error().Msg("can not get `user` session value")
```
Comment (peusebiu) [inline pkg/api/authn.go:835]: same as above

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**